src/lib/formats/flacfile.cpp

```diff
@@ -0,0 +1,87 @@
+// license:BSD-3-Clause
+// copyright-holders:Wilbert Pol
+/*********************************************************************
+
+    flacfile.cpp
+
+    Format code for flac files.
+
+*********************************************************************/
+
+#include "flacfile.h"
+#include "flac.h"
+
+
+static constexpr int MAX_CHANNELS = 8;
+
+
+cassette_image::error flacfile_identify(cassette_image *cassette, cassette_image::Options *opts)
+{
+	const uint64_t file_size = cassette->image_size();
+	std::vector<uint8_t> file_contents(file_size);
+	cassette->image_read(&file_contents[0], 0, file_size);
+	flac_decoder decoder(&file_contents[0], file_size);
+	decoder.reset();
+	const int channels = decoder.channels();
+	const int sample_rate = decoder.sample_rate();
+	const int bits_per_sample = decoder.bits_per_sample();
+	const int total_samples = decoder.total_samples();
+	decoder.finish();
+
+	opts->channels = channels;
+	opts->sample_frequency = sample_rate;
+	opts->bits_per_sample = bits_per_sample;
+
+	if (channels > MAX_CHANNELS)
+	{
+		return cassette_image::error::INVALID_IMAGE;
+	}
+	else if (channels > 0 && sample_rate > 0 && total_samples > 0)
+	{
+		return cassette_image::error::SUCCESS;
+	}
+	else
+	{
+		return cassette_image::error::INVALID_IMAGE;
+	}
+}
+
+
+static cassette_image::error flacfile_load(cassette_image *cassette)
+{
+	const uint64_t file_size = cassette->image_size();
+	std::vector<uint8_t> file_contents(file_size);
+	cassette->image_read(&file_contents[0], 0, file_size);
+	flac_decoder decoder(&file_contents[0], file_size);
+	decoder.reset();
+	const int channels = decoder.channels();
+	const int total_samples = decoder.total_samples();
+
+	std::unique_ptr<int16_t[]> samples[MAX_CHANNELS];
+	int16_t* channel_samples[MAX_CHANNELS];
+	for (int channel = 0; channel < channels; channel++)
+	{
+		samples[channel] = std::make_unique<int16_t[]>(total_samples);
+		channel_samples[channel] = &samples[channel][0];
+	}
+	decoder.decode(channel_samples, decoder.total_samples(), false);
+	for (int channel = 0; channel < channels; channel++)
+	{
+		cassette_image::error err = cassette->put_samples(channel, 0.0,
+			(double)total_samples/decoder.sample_rate(), total_samples, 2,
+			channel_samples[channel], cassette_image::WAVEFORM_16BITLE);
```

Comment:
Does the decoder produce little Endian samples irrespective of host Endianness, or is it native word format?  I’d rather avoid introducing more Endianness bugs.

Reply:
Good point. Looking at the code in flac.cpp, the flac library code produces data in host endianness. I have changed the decode call to swap endianness when the host is big endian so the generated samples should always be little endian.